src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -651,55 +651,48 @@ private void RunOnce(bool checkConditions)
             while (scheduledTasks.Count > 0)
             {
                 // Run all tasks until empty
-                RunAllTasks();
-
-                // If there are any conditions, schedule a new condition-check task and the run all
-                // tasks again. This is a scheduled task instead of just run inline because it needs
-                // to be in the workflow context (i.e. current task scheduler) so the `Workflow`
-                // methods work properly.
-                if (checkConditions && conditions.Count > 0)
+                while (scheduledTasks.Count > 0)
                 {
-                    _ = QueueNewTaskAsync(CheckConditionsAsync);
-                    RunAllTasks();
-                }
-            }
-        }
-
-        private void RunAllTasks()
-        {
-            while (scheduledTasks.Count > 0)
-            {
-                // Pop last
-                var task = scheduledTasks.Last!.Value;
-                scheduledTasks.RemoveLast();
-                scheduledTaskNodes.Remove(task);
+                    // Pop last
+                    var task = scheduledTasks.Last!.Value;
+                    scheduledTasks.RemoveLast();
+                    scheduledTaskNodes.Remove(task);
 
-                // This should never return false
-                if (!TryExecuteTask(task))
-                {
-                    logger.LogWarning("Task unexpectedly was unable to execute");
-                }
-                if (currentActivationException != null)
-                {
-                    ExceptionDispatchInfo.Capture(currentActivationException).Throw();
+                    // This should never return false
+                    if (!TryExecuteTask(task))
+                    {
+                        logger.LogWarning("Task unexpectedly was unable to execute");
+                    }
+                    if (currentActivationException != null)
+                    {
+                        ExceptionDispatchInfo.Capture(currentActivationException).Throw();
+                    }
                 }
-            }
-        }
 
-        private Task CheckConditionsAsync()
-        {
-            try
-            {
-                foreach (var source in conditions.Where(t => t.Item1()).Select(t => t.Item2))
+                // Check conditions. It would be nice if we could run this in the task scheduler
+                // because then users could have access to the `Workflow` context in the condition
+                // callback. However, this cannot be done because even just running one task in the
+                // scheduler causes .NET to add more tasks to the scheduler. And you don't want to
```

Comment:
What other tasks is dotnet adding ?

Reply:
Just looked at stack trace, it seems like this is actually in my stack trace logic at https://github.com/temporalio/sdk-dotnet/blob/1f6676bb9563ddb8eb4606a8df650fb540420fc2/src/Temporalio/Worker/WorkflowInstance.cs#L730 (and just below it for the return-type-based one) which is occurring during stack trace test (stack trace disabled by default in .NET). So I could technically disable stack trace continuations during condition checking, but I think the behavior here in the PR of not pushing the condition checks into the task queue is the best anyways.